devops-970.nix

```diff
@@ -0,0 +1,144 @@
+
+let
+  old = import ./. { inherit cluster; };
+  inherit (old) pkgs;
+  inherit (pkgs) lib;
+  newpkgssrc = pkgs.fetchFromGitHub {
+    owner = "nixos";
+    repo = "nixpkgs";
+    rev = "be445a9074f";
+    sha256 = "15dc7gdspimavcwyw9nif4s59v79gk18rwsafylffs9m1ld2dxwa";
+  };
+  newpkgs = import newpkgssrc {
+    #crossSystem = (import "${newpkgssrc}/lib").systems.examples.mingwW64;
+    #localSystem.system = "x86_64-linux";
+  };
+  ghcWithCardano = old.temp.haskellPackages.ghcWithPackages (ps: [ ps.cardano-sl ps.cardano-sl-x509 ]);
+  packages = self: {
+    inherit old;
+    dlls = pkgs.fetchurl {
+      url = "https://s3.eu-central-1.amazonaws.com/daedalus-ci-binaries/DLLs.zip";
+      sha256 = "0p6nrf8sg2wgcaf3b1qkbb98bz2dimb7lnshsa93xnmia9m2vsxa";
+    };
+    configMutator = pkgs.runCommand "configMutator" { buildInputs = [ ghcWithCardano ]; } ''
+      cp ${./ConfigMutator.hs} ConfigMutator.hs
+      mkdir -p $out/bin/
+      ghc ConfigMutator.hs -o $out/bin/config-mutator
+    '';
+    nsis = newpkgs.callPackage ./nsis.nix {};
+    cardanoJSON = builtins.fromJSON (builtins.readFile ./cardano-sl-src.json);
+    cardanoSrc = pkgs.fetchFromGitHub {
+      owner = "input-output-hk";
+      repo = "cardano-sl";
+      rev = self.cardanoJSON.rev;
+      sha256 = self.cardanoJSON.sha256;
+    };
+    cross = (import (self.cardanoSrc + "/release.nix") { cardano = { outPath = self.cardanoSrc; rev = self.cardanoJSON.rev; }; }).daedalus-mingw32-pkg;
+    unpackedCardano = pkgs.runCommand "daedalus-bridge" { buildInputs = [ pkgs.unzip ]; } ''
+      mkdir $out
+      cd $out
+      unzip ${self.cross}/CardanoSL.zip
+    '';
+    signFile = file: let
+      signingScript = pkgs.writeScript "signing-script" ''
+        #!${pkgs.stdenv.shell}
+
+        exec 3>&1
+        exec 1>&2
+
+        export PATH=${pkgs.mono}/bin:$PATH
+        PASS=hunter2
+
+        DIR=$(realpath $(mktemp -d))
+        cd $DIR
+        cp ${file} .
+        FILE=$(basename ${file})
+        chmod +w $FILE
+
+        # if stdout is a tty, then mono 5.8 will barf over the terminfo files being too new
+        # mono 5.16 supports them, but isnt in our current nixpkgs
+        # for more info, refer to `mcs/class/corlib/System/TermInfoReader.cs` and `ReadHeader`
+        echo $PASS | signcode -spc ${toString signingKeys.spc} -v ${toString signingKeys.pvk} -a sha1 -$ commercial -n "TODO description" -i http://iohk.io -t http://timestamp.verisign.com/scripts/timstamp.dll -tr 10 $FILE | cat
+        storePath=$(nix-store --add-fixed sha256 $FILE)
+        rm -rf $DIR
+        echo $storePath >&3
+      '';
+      res = builtins.exec [ signingScript ];
+    in res;
+    signedCardano = pkgs.runCommand "signed-daedalus-bridge" {} ''
+      cp -r ${self.unpackedCardano} $out
+      chmod -R +w $out
+      cd $out
+      rm *.exe
+      cp ${self.signFile "${self.unpackedCardano}/cardano-launcher.exe"} cardano-launcher.exe
```

Comment:
Note for future generations: this is is effectively _import-from-derivation_, because `signFile` is calls `unsafePerformExec`.

Reply:
yep, `builtins.exec` runs at eval time, so all of its inputs (the `unpackedCardano`) must be built at eval time